File: 백준/Gold/3673. 나눌 수 있는 부분 수열/나눌 수 있는 부분 수열.py
import sys; input = sys.stdin.readline
T = int(input())

for _ in range(T):
    division, N = map(int, input().split())
    arr = [0] + list(map(int, input().split()))
    len_arr = len(arr)
    prefix = [0 for _ in range(N+1)]
    nameoji = {}
    ans = 0
    for i in range(1, len_arr):
        prefix[i] = prefix[i-1] + arr[i]
        idx = prefix[i]%division
        nameoji[idx] = nameoji.get(idx, 0) + 1

    for key in nameoji.keys():
        ans += nameoji[key]*(nameoji[key]-1)//2

    ans += nameoji.get(0, 0)
    print(ans)

File: 백준/Gold/3673. 나눌 수 있는 부분 수열/나눌 수 있는 부분 수열.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Map;
import java.util.StringTokenizer;
import java.util.TreeMap;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		int c = Integer.parseInt(st.nextToken());
		StringBuilder sb = new StringBuilder();
		while (c-- > 0) {
			st = new StringTokenizer(br.readLine());
			int d = Integer.parseInt(st.nextToken());
			int n = Integer.parseInt(st.nextToken());
			int[] arr = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
			int[] prefix = new int[n+1];
			for (int i=0; i<n; i++) {
				prefix[i + 1] = (prefix[i] + arr[i]) % d;
			}
			long ans = 0;
			Map<Integer, Long> map = new TreeMap<>();
			for (int i=0; i<=n; i++) {
				ans += map.getOrDefault(prefix[i], Long.valueOf(0));
				map.put(prefix[i], map.getOrDefault(prefix[i], Long.valueOf(0)) + 1);
			}
			sb.append(ans).append("\n");
		}
		System.out.print(sb);
	}
}